src/vr/journal.zig

```diff
@@ -11,18 +11,87 @@ const Storage = @import("../storage.zig").Storage;
 
 const vr = @import("../vr.zig");
 const Header = vr.Header;
-
-const ConcurrentRanges = @import("../concurrent_ranges.zig").ConcurrentRanges;
-const Range = @import("../concurrent_ranges.zig").Range;
+const Replica = vr.Replica;
 
 pub const Journal = struct {
+    pub const Read = struct {
+        replica: *Replica,
+        completion: Storage.Read,
+        callback: fn (self: *Replica, prepare: ?*Message, destination_replica: ?u16) void,
+
+        message: *Message,
+        op: u64,
+        checksum: u128,
+        destination_replica: ?u16,
+    };
+
+    pub const Write = struct {
+        pub const Trigger = enum { append, repair };
+
+        replica: *Replica,
+        callback: fn (self: *Replica, wrote: ?*Message, trigger: Trigger) void,
+
+        message: *Message,
+        trigger: Trigger,
+
+        /// True if this Write has acquired a lock on a sector of headers.
+        /// This also means that the Write is currently writing sectors or queuing to do so.
+        header_sector_locked: bool = false,
+        /// Linked list of Writes waiting to acquire the same logical header range as this Write.
+        header_sector_next: ?*Write = null,
+
+        /// This is reset to undefined and reused for each Storage.write_sectors() call.
+        range: Range,
+
+        const SectorBuffer = *align(config.sector_size) [config.sector_size]u8;
+        fn headers(write: *Journal.Write, journal: *Journal) SectorBuffer {
+            const i = (@ptrToInt(write) - @ptrToInt(&journal.writes.items)) / @sizeOf(Journal.Write);
```

Comment:
This already asserts that the `write` pointer is `>= items` since if that condition was false we'd get safety-checked underflow. Switching to `@divExact()` will assert that the `write` pointer is a multiple of the type, which is a good idea.

I'll add an assert for the second case you mention.

Reply:
Actually, we already assert that second case as well, since `journal.header_iops.len == journal.writes.items.len` and the index into `header_iops` is bounds-checked. I'll instead add an assert to ensure those lengths match.